### PR TITLE
useReducer

### PR DESCRIPTION
**Pull request by msyleung**
- updated useReducer page
  - added basic description 
  - reducer, dispatch, action definition
  - added third argument lazy initializer explanation
  - added two code samples

### CODE REVIEW

File: examples/files/hooks/useReducer.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import React, { useReducer } from 'react'
+
+const App = () => {
+  const reducer = (state, action) => {
```
Comment (dabbott) [inline examples/files/hooks/useReducer.js:4]: Should the reducer function go outside of `App`? This is going to declare a new function every render, and I'm not actually sure what the behavior of `useReducer` is in that case - but if we move it outside we won't have to consider that case.

File: examples/files/hooks/useReducer.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  }
+  const [state, dispatch] = useReducer(reducer, 'No Pet Selected')
+  return (
+    <div>
+      <input
+        type="radio"
+        name="pet"
+        value="cat"
+        onClick={() => {
+          dispatch({ type: 'cat' })
```
Comment (dabbott) [inline examples/files/hooks/useReducer.js:20]: I think "cat" or "dog" would be the `value` or `payload` of an action, rather than the `type`.

This example is tricky since we want to keep it short, but also, `useReducer` is really only relevant when we have sub-values in our state.

I'd probably make state contain 2 different sub-values, e.g. `{ pet: "cat", color: "black" }`, and then dispatch actions like `{ type: "pet", value: "dog" }`.

We can possibly use `<select>` instead of radio buttons to save space.

It'd come out a bit longer, but, it'd be a more realistic example.

File: pages/hooks/usereducer.mdx [hunk 3]
```diff
@@ -2,6 +2,78 @@
...
+useReducer is similar to useState since it allows us to control state inside a function component.  
+The difference between the two is that useReducer allows us to handle multiple states and control complex state changes. 
+
+The useReducer hook takes two mandatory arguments: a reducer function and an initial argument. It can also take in an optional third argument called the initializer, which we will discuss later. useReducer returns the current state and a dispatch function.  
+
+`const [state, dispatch] = useReducer(reducer, initialState)`
+
+## Reducer
+
+The reducer function is a pure function that takes in the previous state and an action. It returns a new state.
```
Comment (dabbott) [inline pages/hooks/usereducer.mdx:20]: I would mention that `action` is an arbitrary value we choose that can be any type

File: pages/hooks/usereducer.mdx [hunk 4]
```diff
@@ -2,6 +2,78 @@
...
+
+The dispatch function, when called, passes the action to the reducer.
+
+In the Example code below, the initial state is `No Pet Selected`. When we click the `cat` radio button, dispatch is called with `{type: 'cat'}` as its argument. This argument gets passed to the reducer as `action`. Then the reducer follows the logic `if (action.type === 'cat')` and returns the new state `You selected cat`
+
+<Example code={useReducer}/>
+
+## useState and useReducer
+
+Here we will highlight some of the differences between useState and useReducer.
```
Comment (dabbott) [inline pages/hooks/usereducer.mdx:40]: I agree that the Refactor example is a bit long for how much value it adds to the topic.

I think it might be enough to explain the difference without a new example - e.g. we can compare and contrast useReducer + `{ pet, color }` vs. 2 separate useState + `pet` and `color`

File: pages/hooks/usereducer.mdx [hunk 5]
```diff
@@ -2,6 +2,78 @@
...
+
+There are situations when one hook is the preferable choice over the other. 
+
+For simpler state transition that use JavaScript primitives, it would be ideal to use useState. 
+
+When there are different properties that are dependent on one another, JavaScript objects or arrays as state, or complex state business logic, it is recommended to useReducer because state will be managed in one place and it would provide a more predictable state architecture. 
+
+## The third argument: Initializer
+
+useReducer allows us to initialize state in two different ways.
```
Comment (dabbott) [inline pages/hooks/usereducer.mdx:55]: I think if we want to show an example for the initializer arg, we need to think of an example with a computationally expensive function, since if people were to copy this example directly as a pattern, it wouldn't be the best.

If we can't think of a good one, we could probably get away with just mentioning the 3rd arg and linking to the react docs.

File: pages/hooks/usereducer.mdx [hunk 6]
```diff
@@ -2,6 +2,78 @@
...
+Recall our example in useEffect, there were multiple declarations of useState and several areas where state was changed.  
+We can refactor the code with `useReducer` and handle all state changes in one `reducer` function. 
+
+<Example code={useReducerRefactor}/>
+
+There are situations when one hook is the preferable choice over the other. 
+
+For simpler state transition that use JavaScript primitives, it would be ideal to use useState. 
+
+When there are different properties that are dependent on one another, JavaScript objects or arrays as state, or complex state business logic, it is recommended to useReducer because state will be managed in one place and it would provide a more predictable state architecture.
```
Comment (dabbott) [inline pages/hooks/usereducer.mdx:51]: I think we need to either give a definition of **predictable state architecture**, or explain this concept in simpler terms.

File: pages/hooks/usereducer.mdx (hunk 5 above)
Comment (msyleung) [reply pages/hooks/usereducer.mdx:55]: Sounds good to me! 👍

File: pages/hooks/usereducer.mdx (hunk 6 above)
Comment (msyleung) [reply pages/hooks/usereducer.mdx]: Not sure if my update made it any better! 😬 I changed it to: 
```
Use `useReducer` when there are different properties that are dependent on one another, 
JavaScript objects or arrays as state, or complex state business logic.   
The component state will be managed in one place and it would be easier to test and debug.
```
